### DISCUSSION TITLE
How to update content on a smartObjectLayer when process PSD with multi layers

### DISCUSSION THREAD

**Discussion by dvcuong05** [dlemstra/Magick.NET#935] [Help]
Hi @all,

I have a requirement to generate a batch of .png files from a single .PSD template file by replacing the new image and background.

1./ Now based on what I researching about MagickImage I can load  .PSD file into MagickImageCollection which contains one of my layers with the label is "PTS1" for example (that is SmartObjectLayer with containing is a picture). So my concern is how to update/replace contain the picture on that layer with another content then export the final .PSD into PNG.?

2/ The same question but for the "SolidLayer" which is fillup the color background of my PSD file. I want to change from color A to B then export PSD to PNG?

Sample data: [sample PSD file.zip](https://github.com/dlemstra/Magick.NET/files/6528227/sample.PSD.file.zip)
Here my code but the result is not as expected:
`public static void processPSD(string psdFile, string newImagePath)
        {

            using (MagickImageCollection magickImages = new MagickImageCollection(psdFile))
            {
                for (var i = 0; i < magickImages.Count; i++)
                {
                    if (magickImages[i].Label == "PTS1")
                    {
                        // replace image content by another.
                       MagickImage newImageContent = new MagickImage(newImagePath);
                       magickImages[i] = newImageContent;
                    }

                    if (magickImages[i].Label == "COLOR")
                    {
                        magickImages[i].BackgroundColor = MagickColors.Red;
                    }
                }
                magickImages.Write("output.psd", MagickFormat.Psd);
                //magickImages[0].Write("output.png", MagickFormat.Png);
            }
        }`

- Input PSD looks like:
![image](https://user-images.githubusercontent.com/4995777/119263693-519ef980-bc0a-11eb-84e1-d4ea4a7d5df1.png)
- Result new PSD:
![image](https://user-images.githubusercontent.com/4995777/119263894-21a42600-bc0b-11eb-8da1-c380f0824b03.png)
- Result in .PNG => no changes are persist it seems, because my new image content should be the music band
![image](https://user-images.githubusercontent.com/4995777/119263957-64fe9480-bc0b-11eb-9421-89371fa259b5.png)

Thanks for any support.

**Comment by dlemstra** [comment]
Updating the text of a PSD file is not possible. The ImageMagick library will create an image from the PSD file and that means all vector information is gone. I have no idea which library is capable of doing this.

**Reply by dvcuong05** [reply]
Thanks @dlemstra 
But my concern question above how to update Image content on a specific layer which is SmartObjectLayer Or Does it the same with text? so I not able to update anything on the layer of PDF file.

**Reply by dlemstra** [reply]
When ImageMagick reads the PSD file it will change the information of the file into pixels. This means that you cannot change the content any objecs defined in Photoshop.
